File: neet_code_150_cpp/binary_search/bin_search.cpp
#include <vector>
#include <iostream>

class Solution {
    public:
        int search(std::vector<int> &nums, int target) {
            int left = 0, right = nums.size() - 1;
            while(left <= right) {
                int mid = (left + right) / 2;
                if(nums[mid] == target) {
                    return mid;
                }

                if(nums[mid] < target) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }

            return -1;
        }

        int search_size_t(std::vector<int> &nums, int target) {
            std::size_t left = 0, right = nums.size();

            while(left < right) {
                std::size_t mid = (right + left) / 2;

                if(nums[mid] == target) {
                    return mid;
                }

                if(nums[mid] < target) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }

            return -1;
        }
};

int main(void) {
    Solution s_1;

    auto case_1_nums = std::vector{-1, 0, 3, 5, 9, 12};
    auto case_1_target = 9;

    auto case_2_nums = std::vector{-1, 0, 3, 5, 9, 12};
    auto case_2_target = 2;

    auto case_3_nums = std::vector{2, 5};
    auto case_3_target = 0;

    std::cout << "index: int" << std::endl;
    std::cout << s_1.search(case_1_nums, case_1_target) << std::endl;
    std::cout << s_1.search(case_2_nums, case_2_target) << std::endl;
    std::cout << s_1.search(case_3_nums, case_3_target) << std::endl;

    std::cout << "index: std::size_t" << std::endl;
    std::cout << s_1.search_size_t(case_1_nums, case_1_target) << std::endl;
    std::cout << s_1.search_size_t(case_2_nums, case_2_target) << std::endl;
    std::cout << s_1.search_size_t(case_3_nums, case_3_target) << std::endl;

}

File: neet_code_150/src/binary_search/bin_search.rs
struct Solution {}
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut left, mut right) = (0, nums.len() as i32 - 1);

        while left <= right {
            let mid = (left + right) / 2;
            if nums[mid as usize] == target {
                return mid as i32;
            }

            if nums[mid as usize] < target {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        -1
    }
}

// 模範解答
use std::cmp::Ordering::*;
struct SolutionAns {}
impl SolutionAns {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut l, mut r) = (0, nums.len());

        while l < r {
            let m = l + (r - l) / 2;
            match target.cmp(&nums[m]) {
                Equal => return m as i32,
                Less => r = m,
                Greater => l = m + 1,
            }
        }

        -1
    }
}

fn main() {
    let case_1 = (vec![-1, 0, 3, 5, 9, 12], 9);
    let case_2 = (vec![-1, 0, 3, 5, 9, 12], 2);
    let case_3 = (vec![2, 5], 0);

    println!("case_1: {}", Solution::search(case_1.0.clone(), case_1.1));
    println!("case_2: {}", Solution::search(case_2.0.clone(), case_2.1));
    println!("case_3: {}", Solution::search(case_3.0.clone(), case_3.1));

    println!("case_1: {}", SolutionAns::search(case_1.0.clone(), case_1.1));
    println!("case_2: {}", SolutionAns::search(case_2.0.clone(), case_2.1));
    println!("case_3: {}", SolutionAns::search(case_3.0.clone(), case_3.1));
}
